sonarr/command_test.go

```diff
@@ -1,35 +1,277 @@
+		withError        error
+		withRequest      *sonarr.CommandRequest
+		expectedRequest  string
+		expectedMethod   string
+		expectedResponse *sonarr.CommandResponse
+	}{
+		{
+			name:           "200",
+			expectedPath:   path.Join("/", starr.API, sonarr.APIver, "command"),
+			responseStatus: http.StatusOK,
+			responseBody: `{"id":1234,"name":"SomeCommand","commandName":"SomeCommandName","message":` +
+				`"Command Message","priority":"testalert","status":"statusalert","queued":` + string(datejson) +
+				`,"started":` + string(datejson) + `,"ended":` + string(datejson) +
+				`,"stateChangeTime":` + string(datejson) + `,"lastExecutionTime":` + string(datejson) +
+				`,"duration":"woofun","trigger":"someTrigger","sendUpdatesToClient":true,"updateScheduledTask":true` +
+				`,"body": {"mapstring": "mapinterface"}` +
+				`}`,
+			withError: nil,
+			withRequest: &sonarr.CommandRequest{
+				Name:      "SomeCommand",
+				SeriesIDs: []int64{1, 3, 7},
+			},
+			expectedRequest: `{"name":"SomeCommand","seriesIds":[1,3,7]}` + "\n",
+			expectedMethod:  "POST",
+			expectedResponse: &sonarr.CommandResponse{
+				ID:                  1234,
+				Name:                "SomeCommand",
+				CommandName:         "SomeCommandName",
+				Message:             "Command Message",
+				Priority:            "testalert",
+				Status:              "statusalert",
+				Queued:              somedate,
+				Started:             somedate,
+				Ended:               somedate,
+				StateChangeTime:     somedate,
+				LastExecutionTime:   somedate,
+				Duration:            "woofun",
+				Trigger:             "someTrigger",
+				SendUpdatesToClient: true,
+				UpdateScheduledTask: true,
+				Body:                map[string]interface{}{"mapstring": "mapinterface"},
+			},
+		},
+		{
+			name:             "404",
+			expectedPath:     path.Join("/", starr.API, sonarr.APIver, "command"),
+			responseStatus:   http.StatusNotFound,
+			responseBody:     `{"message": "NotFound"}`,
+			withError:        starr.ErrInvalidStatusCode,
+			expectedMethod:   "POST",
+			expectedResponse: nil,
+			withRequest:      &sonarr.CommandRequest{Name: "Something"},
+			expectedRequest:  `{"name":"Something"}` + "\n",
+		},
+		{
+			name:             "noname", // no name provided? returns empty (non-nil) response.
+			withRequest:      &sonarr.CommandRequest{Name: ""},
+			expectedResponse: &sonarr.CommandResponse{},
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				assert.Equal(t, test.expectedPath, req.URL.String())
+				w.WriteHeader(test.responseStatus)
+
+				_, err := w.Write([]byte(test.responseBody))
+				assert.NoError(t, err)
+				assert.Equal(t, req.Method, test.expectedMethod)
+
+				body, err := ioutil.ReadAll(req.Body)
+				assert.NoError(t, err)
+				assert.Equal(t, test.expectedRequest, string(body))
+			}))
```

Comment:
Nice that you check the method as well! maybe the request body as a string is even too much.

Just one thought that came to my mind today about this:
If you want, we can even define this function as a reusable generic helper, maybe in sonarr_test.go. something like:
```go
func initMockServer(t *testing.T, test interface{}) *httptest.Server {
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		assert.Equal(t, test.expectedPath, req.URL.String())
		w.WriteHeader(test.responseStatus)

		_, err := w.Write([]byte(test.responseBody))
		assert.NoError(t, err)
		assert.Equal(t, req.Method, test.expectedMethod)

		body, err := ioutil.ReadAll(req.Body)
		assert.NoError(t, err)
		assert.Equal(t, test.expectedRequest, string(body))
	}
}
```
I don't know if it will work, since tests can be different from each other. Maybe we can define 2 functions, one with body and one without.

Reply:
I thought about this, but I didn't come up with a good way to implement it, since every method has a different output. We could probably just set it to `interface{}` though honestly because `assert.Equal` will figure it out. We should definitely try this.

Reply:
if the interface is `nil` we would assume it's a body-less request? Will have to play with this idea....

Reply:
maybe we can leverage the expected method? if it's GET, then we can probably skip the body check

Reply:
Also a good idea. We use four methods. Two have inputs, two do not. I'll work on this. :)

Reply:
Okay. It's converted to a method-call.

Reply:
I think it's good!